keyring.go

```diff
@@ -348,25 +348,24 @@ func (d *DirectKeyFetcher) FetchKeys(
 }
 
 func (d *DirectKeyFetcher) fetchKeysForServer(
-	ctx context.Context, serverName ServerName, requests map[PublicKeyRequest]Timestamp,
+	ctx context.Context, serverName ServerName,
 ) (map[PublicKeyRequest]PublicKeyLookupResult, error) {
-	serverKeys, err := d.Client.LookupServerKeys(ctx, serverName, requests)
+	keys, err := d.Client.GetServerKeys(ctx, serverName)
 	if err != nil {
 		return nil, err
 	}
+	// Check that the keys are valid for the server.
+	checks, _, _ := CheckKeys(serverName, time.Unix(0, 0), keys, nil)
+	if !checks.AllChecksOK {
+		return nil, fmt.Errorf("gomatrixserverlib: key response direct from %q failed checks", serverName)
+	}
 
 	results := map[PublicKeyRequest]PublicKeyLookupResult{}
-	for _, keys := range serverKeys {
-		// Check that the keys are valid for the server.
-		checks, _, _ := CheckKeys(serverName, time.Unix(0, 0), keys, nil)
-		if !checks.AllChecksOK {
-			return nil, fmt.Errorf("gomatrixserverlib: key response direct from %q failed checks", serverName)
-		}
 
-		// TODO: What happens if the same key ID appears in multiple responses?
-		// We should probably take the response with the highest valid_until_ts.
-		mapServerKeysToPublicKeyLookupResult(keys, results)
-	}
+	// TODO (matrix-org/dendrite#345): What happens if the same key ID
+	// appears in multiple responses? We should probably take the response
+	// with the highest valid_until_ts.
+	mapServerKeysToPublicKeyLookupResult(keys, results)
```

Comment:
Is this still true if we're only calling this once?

Reply:
yes, the server could still return two keys with the same ID. Though maybe we should reject the whole response in that case?

Reply:
It feels like something has gone wrong somewhere along the lines if that happens, but I don't really mind much what we do.

Reply:
To be fair this comment probably applies more to the perspectives case, which I forgot to update; done in 886000e.